astropy/utils/masked/tests/test_functions.py

```diff
@@ -0,0 +1,320 @@
+        assert_array_equal(ma_reduce.mask, expected_mask)
+
+    def test_ufunc_not_implemented_for_other(self):
+        """
+        If the unmasked operation returns NotImplemented, this
+        should lead to a TypeError also for the masked version.
+        """
+        a = np.array([1, 2])
+        b = 3 * u.m
+        with pytest.raises(TypeError):
+            a & b
+
+        ma = Masked(a)
+        with pytest.raises(TypeError):
+            ma & b
+
+
+class TestMaskedQuantityUfuncs(MaskedUfuncTests, QuantitySetup):
+    def test_ufunc_inplace_error2(self):
+        out = Masked(np.zeros(self.ma.shape))
+        with pytest.raises(TypeError):
+            np.add(self.ma, self.mb, out=out)
+
+
+class TestMaskedLongitudeUfuncs(MaskedUfuncTests, LongitudeSetup):
+    def test_ufunc_inplace_quantity_initial(self):
+        out = Masked(np.zeros(self.ma.shape) << u.m)
+        result = np.add(self.ma, self.mb, out=out)
+        assert result is out
+        expected = np.add(self.ma, self.mb).view(Quantity)
+        assert_masked_equal(result, expected)
+
+
+class TestMaskedArrayConcatenation(MaskedArraySetup):
+    def test_concatenate(self):
+        mb = self.mb[np.newaxis]
+        concat_a_b = np.concatenate((self.ma, mb), axis=0)
+        expected_data = np.concatenate((self.a, self.b[np.newaxis]), axis=0)
+        expected_mask = np.concatenate((self.mask_a, self.mask_b[np.newaxis]),
+                                       axis=0)
+        assert_array_equal(concat_a_b.unmasked, expected_data)
+        assert_array_equal(concat_a_b.mask, expected_mask)
+
+    def test_concatenate_not_all_masked(self):
+        mb = self.mb[np.newaxis]
+        concat_a_b = np.concatenate((self.a, mb), axis=0)
+        expected_data = np.concatenate((self.a, self.b[np.newaxis]), axis=0)
+        expected_mask = np.concatenate((np.zeros(self.a.shape, bool),
+                                        self.mask_b[np.newaxis]), axis=0)
+        assert_array_equal(concat_a_b.unmasked, expected_data)
+        assert_array_equal(concat_a_b.mask, expected_mask)
+
+    @pytest.mark.parametrize('obj', (1, slice(2, 3)))
+    def test_insert(self, obj):
+        mc_in_a = np.insert(self.ma, obj, self.mc, axis=-1)
+        expected = Masked(np.insert(self.a, obj, self.c, axis=-1),
+                          np.insert(self.mask_a, obj, self.mask_c, axis=-1))
+        assert_masked_equal(mc_in_a, expected)
+
+    def test_insert_masked_obj(self):
+        with pytest.raises(TypeError):
+            np.insert(self.ma, Masked(1, mask=False), self.mc, axis=-1)
+
+    def test_append(self):
+        mc_to_a = np.append(self.ma, self.mc, axis=-1)
+        expected = Masked(np.append(self.a, self.c, axis=-1),
+                          np.append(self.mask_a, self.mask_c, axis=-1))
+        assert_masked_equal(mc_to_a, expected)
+
+
+class TestMaskedQuantityConcatenation(TestMaskedArrayConcatenation,
+                                      QuantitySetup):
+    pass
```

Comment:
Needs completion.

Reply:
This one is OK: it just reruns the tests from `TestMaskedArrayConcatenation` with quantities instead of masked arrays (using `QuantitySetup`).

Reply:
Ah, clever.